lib/spoom/deadcode/indexer.rb

```diff
@@ -0,0 +1,394 @@
+        @in_opassign = true
+        super
+        @in_opassign = false
+      end
+
+      sig { params(send: Send).void }
+      def visit_send(send)
+        visit(send.recv)
+
+        case send.name
+        when "attr_reader"
+          send.args.each do |arg|
+            next unless arg.is_a?(SyntaxTree::SymbolLiteral)
+
+            name = symbol_string(arg)
+            define_attr_reader(name, [*@names_nesting, name].join("::"), arg)
+          end
+        when "attr_writer"
+          send.args.each do |arg|
+            next unless arg.is_a?(SyntaxTree::SymbolLiteral)
+
+            name = symbol_string(arg)
+            define_attr_writer("#{name}=", "#{[*@names_nesting, name].join("::")}=", arg)
+          end
+        when "attr_accessor"
+          send.args.each do |arg|
+            next unless arg.is_a?(SyntaxTree::SymbolLiteral)
+
+            name = symbol_string(arg)
+            full_name = [*@names_nesting, name].join("::")
+            define_attr_reader(name, full_name, arg)
+            define_attr_writer("#{name}=", "#{full_name}=", arg)
+          end
+        else
+          reference_method(send.name, send.node)
+          visit_all(send.args)
+          visit(send.block)
+        end
+      end
+
+      sig { override.params(node: SyntaxTree::SymbolLiteral).void }
+      def visit_symbol_literal(node)
+        # Something like `:FOO` will yield a Const node but we do not want to treat it as a constant reference.
+        # So we need to distinguish between the two cases.
+        @in_symbol_literal = true
+        super
+        @in_symbol_literal = false
+      end
+
+      sig { override.params(node: SyntaxTree::TopConstField).void }
+      def visit_top_const_field(node)
+        define_constant(node.constant.value, node.constant.value, node)
+      end
+
+      sig { override.params(node: SyntaxTree::VarField).void }
+      def visit_var_field(node)
+        value = node.value
+        case value
+        when SyntaxTree::Const
+          if @in_opassign
+            reference_constant(value.value, node)
+          else
+            name = value.value
+            define_constant(name, [*@names_nesting, name].join("::"), node)
+          end
+        when SyntaxTree::Ident
+          reference_method(value.value, node) if @in_opassign
+          reference_method("#{value.value}=", node)
+        end
+      end
+
+      sig { override.params(node: SyntaxTree::VCall).void }
+      def visit_vcall(node)
+        visit_send(Send.new(node: node, name: node_string(node.value)))
+      end
+
+      private
+
+      # Definition indexing
+
+      sig { params(name: String, full_name: String, node: SyntaxTree::Node).void }
+      def define_attr_reader(name, full_name, node)
+        definition = Definition.new(
+          kind: Definition::Kind::AttrReader,
+          name: name,
+          full_name: full_name,
+          location: node_location(node),
+        )
+        @index.define(definition)
+      end
```

Comment:
These methods seem to be mostly boilerplate code? Do we expect them to vary in follow up PRs?

Reply:
At this point, indeed and I hesitated to introduce them now. But they will become useful when we introduce plugins as they offer different injection points depending on the kind of nodes you may want to handle from a plugin. Introducing them later was creating a lot of noise in the later PR.

Reply:
Yeah that's what I thought too. Just want to confirm 👍